File: DSA_python/DSA11012 CÂY NHỊ PHÂN BẰNG NHAU.py
from collections import deque

class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def add_to_tree(edges):
    if not edges:
        return

    nodes = dict()

    for parent, child, pos in edges:
        if parent not in nodes:
            nodes[parent] = TreeNode(parent)
        if child not in nodes:
            nodes[child] = TreeNode(child)
        if pos == 'L':
            nodes[parent].left = nodes[child]
        else:
            nodes[parent].right = nodes[child]
        
    children = [child for _,child,_ in edges]
    parents = [parent for parent,_,_ in edges]
    root = parents[0]

    for parent in parents:
        if parent not in children:
            root = parent
    
    return nodes[root]

def traversal(node1, node2):
    if not node1 and not node2:
        return True
    elif (node1.val == node2.val) and node1 and node2:
        return traversal(node1.left, node2.left) and traversal(node1.right, node2.right)
    return False

def submit():
    n = int(input())
    A = list(input().split())
    
    edges = list()
    for i in range(2, len(A), 3):
        edges.append((A[i-2], A[i-1], A[i]))
    root1 = add_to_tree(edges)

    m = int(input())
    B = list(input().split())

    edges = list()
    for i in range(2, len(A), 3):
        edges.append((B[i-2], B[i-1], B[i]))
    root2 = add_to_tree(edges)

    print(1 if traversal(root1, root2) else 0)









test = int(input())
for i in range(test):
    submit()

File: DSA/DSA11012 CÂY NHỊ PHÂN BẰNG NHAU.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};




bool add(TreeNode *node, int parent, int child, char pos)
{
    if (!node) return false;
    
    if (node->val != parent)
    {
        if (add(node->left, parent, child, pos)) return true;
        if (add(node->right, parent, child, pos)) return true;
        return false;
    }
    else
    {
        if (pos == 'L')
        {
            if (!node->left) node->left = new TreeNode(child);
        }
        else
        {
            if (!node->right) node->right = new TreeNode(child);
        }
        return true;
    }
}

bool check_same_tree(TreeNode *root1, TreeNode *root2)
{
    if (!root1 && !root2)
    {
        return true;
    }
    else if ((root1->val == root2->val) && root1 && root2)
    {
        return check_same_tree(root1->left, root2->left) && check_same_tree(root1->right, root2->right);
    }
    return false;
}

void submit()
{
    int n;
    int parent, child;
    char pos;
    cin >> n >> parent >> child >> pos;
    
    TreeNode *root1 = new TreeNode(parent);

    add(root1, parent, child, pos);
    
    for (int i = 1; i < n; i++)
    {
        cin >> parent >> child >> pos;
        add(root1, parent, child, pos);
    }



    cin >> n >> parent >> child >> pos;
    
    TreeNode *root2 = new TreeNode(parent);

    add(root2, parent, child, pos);

    for (int i = 1; i < n; i++)
    {
        cin >> parent >> child >> pos;
        add(root2, parent, child, pos);
    }

    check_same_tree(root1, root2) ? cout << 1 : cout << 0;

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}